### PR TITLE
Mrc 5661- FE Run page - packet groups + params

### PR DESCRIPTION
**Pull request by absternator**
Two* fields are added to the run page, as shown in the screenshot. The first is selecting the packet group name. After the user has selected, parameters can be chosen if there are parameters available for the packet group. 

For users to enter params it will be a text box and users will write as strings. And the value gets converted to number, string, boolean depending on what is typed.

Testing: 
1. click on a packet group without params. the params would not show.
2. click on packet group with params (parameters, test1) and parameters should show. play around with typing in different values and clicking run. if you look at console this is what will be sent to server when the submit functionality is done in next pr.

image

### CODE REVIEW

File: app/src/app/components/contents/runner/run/PacketRunPacketGroupField.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                <CommandList>
+                  <CommandEmpty>No packet groups found.</CommandEmpty>
+                  <CommandGroup>
+                    {packetGroups.map((packetGroupName) => (
+                      <CommandItem
+                        value={packetGroupName.name}
+                        key={packetGroupName.name}
+                        onSelect={() => {
+                          form.setValue("packetGroupName", packetGroupName.name);
+                        }}
```
Comment (EmmaLRussell) [inline app/src/app/components/contents/runner/run/PacketRunPacketGroupField.tsx:49]: One usability issue on packet group select - the drop down doesn't immediately close when you select a value (on Firefox). Need to click outside the control to get it to close.

**Review by EmmaLRussell — COMMENTED** [review]
Looks good to me. One query about the parameter metadata - I feel like we should have parameter type and defaults available, but maybe I'm misremembering, Also a couple of small questions just related to my ongoing confusion about React forma! :laughing: 

> For users to enter params it will be a text box and users will write as strings. And the value gets converted to number, string, boolean depending on what is typed.

Does the packet group not advertise what parameter types each of its parameters supports, and so could we not make the input UI appropriate to that type i.e. text input, numeric input, checkbox? It seems a bit error-prone to make people type "true" for example.

**Comment by absternator** [comment]
> Does the packet group not advertise what parameter types each of its parameters supports, and so could we not make the input UI appropriate to that type i.e. text input, numeric input, checkbox? It seems a bit error-prone to make people type "true" for example.

Yup I fully agree that would be much better.. the issue is we don't get the type back so cant do that :cry: ... we only get the default value or null if no default

File: app/src/app/components/contents/runner/run/PacketRunForm.tsx [hunk 2]
```diff
@@ -2,38 +2,66 @@ import { zodResolver } from "@hookform/resolvers/zod";
...
-  parameters: z.record(z.any())
+  branch: z.string({ required_error: "Branch is required" }),
+  packetGroupName: z.string({ required_error: "Packet Group name is required" }),
+  parameters: z.array(
+    z.object({
+      name: z.string(),
+      value: z
+        .string()
+        .min(1, "Must enter a number, string or boolean")
+        .refine((value) => value !== "null", { message: "Value cannot be null." })
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/src/app/components/contents/runner/run/PacketRunForm.tsx:26]: no this is if the user literally types in 'null'... if the user leaves empty form wont submit

File: app/src/app/components/contents/runner/run/PacketRunPacketGroupFields.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    return <ErrorComponent message="Error loading packet groups" error={error} />;
+  }
+  if (isLoading) {
+    return <Skeleton className="h-12 w-96" />;
+  }
+
+  return packetGroups ? (
+    <>
+      <PacketRunPacketGroupField form={form} packetGroups={packetGroups} />
+      <PacketRunParamsField branchCommit={branchCommit} form={form} packetGroupName={form.watch("packetGroupName")} />
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/src/app/components/contents/runner/run/PacketRunPacketGroupFields.tsx:27]: yup exactly

File: app/src/app/components/contents/runner/run/PacketRunParamsField.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          {fields.length == 0 ? (
+            <FormDescription>No parameters available for this packet group.</FormDescription>
+          ) : (
+            <>
+              <FormDescription>Fill in parameters with a number, string or boolean</FormDescription>
+              {fields.map((item, idx) => (
+                <FormField
+                  key={item.id}
+                  control={form.control}
+                  name={`parameters.${idx}.value`}
```
[the remark this answers is not in the document]
Comment (absternator) [reply app/src/app/components/contents/runner/run/PacketRunParamsField.tsx:52]: yup so the name must map to form schema and it knows to recognize it...

File: app/src/app/components/contents/runner/run/PacketRunPacketGroupField.tsx (hunk 1 above)
Comment (absternator) [reply app/src/app/components/contents/runner/run/PacketRunPacketGroupField.tsx]: agreed have made it close on select

File: app/src/app/components/contents/runner/utils/parseParameterValue.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+export const parseParameterValue = (value: string): string | number | boolean | null => {
+  switch (true) {
+    case value === "":
+      return null;
```
Comment (M-Kusumgar) [inline app/src/app/components/contents/runner/utils/parseParameterValue.ts:4]: Wait I am a bit confused, the form won't submit if any of the values are empty right? Why do we have this case?

**Review by M-Kusumgar — COMMENTED** [review]
Looks awesome, just got a small query for you. Also @EmmaLRussell the orderly function lets users define parameters as `a = NULL` in R, so there's no way to get any information about what type of thing that want to pass into the parameter `a`, we could add a type arg or something but not every report will be guaranteed to have that and I do feel like we should make the web fit orderly including R's miserable typing 🤣🤣

Btw suggestion for future: it would be really good to have a reset to default param value if it exists button for the parameters!

File: app/src/app/components/contents/runner/utils/parseParameterValue.ts (hunk 5 above)
Comment (absternator) [reply app/src/app/components/contents/runner/utils/parseParameterValue.ts:4]: you are right and this was fixed as part of next PR #116

**Review by M-Kusumgar — APPROVED** [review]

**Comment by EmmaLRussell** [comment]
> Looks awesome, just got a small query for you. Also @EmmaLRussell the orderly function lets users define parameters as `a = NULL` in R, so there's no way to get any information about what type of thing that want to pass into the parameter `a`, we could add a type arg or something but not every report will be guaranteed to have that and I do feel like we should make the web fit orderly including R's miserable typing 🤣🤣
> 
In that case I think we need to do more in the front end to make sure the user ends up with the types they expect.

File: app/src/app/components/contents/runner/run/PacketRunForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (EmmaLRussell) [reply app/src/app/components/contents/runner/run/PacketRunForm.tsx:26]: But it's case-sensitive? A user familiar with R would be more likely to type NULL wouldn't they? Which seems to be accepted ok..
And if that is allowed through, the logic would just interpret it as the string "NULL" right - but we just don't want users thinking they can provide null parameters when they can't?

**Review by EmmaLRussell — COMMENTED** [review]
guess I missed the boat on this one 😢

File: app/src/app/components/contents/runner/run/PacketRunForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (absternator) [reply app/src/app/components/contents/runner/run/PacketRunForm.tsx:26]: Oh yea true will update as small or with this

**Comment by absternator** [comment]
> guess I missed the boat on this one 😢 
> 
> 

Sorry I restocked all comments... but saw you added more so will make small pr for it

**Pull request merged**